File: sdk/crewlayer/_memory.py
"""Memory resource clients — sync and async."""
from __future__ import annotations

from typing import Any

from crewlayer._http import AsyncTransport, SyncTransport
from crewlayer._types import (
    ExtractResult,
    MemoryItem,
    MemoryPage,
    RecallResult,
    ShortMemory,
)


class MemoryClient:
    """Synchronous memory operations."""

    def __init__(self, http: SyncTransport) -> None:
        self._http = http

    def append(
        self,
        agent_id: str,
        role: str,
        content: str,
        *,
        session_id: str = "default",
        metadata: dict[str, Any] | None = None,
    ) -> ShortMemory:
        """Append a message to the agent's short-term session memory."""
        data = self._http.request(
            "POST",
            f"/v1/agents/{agent_id}/memory/messages",
            params={"session_id": session_id},
            json={"role": role, "content": content, "metadata": metadata or {}},
        )
        return ShortMemory._from(data)

    def messages(
        self,
        agent_id: str,
        *,
        session_id: str = "default",
        limit: int = 50,
    ) -> ShortMemory:
        """Retrieve recent messages from the agent's session memory."""
        data = self._http.request(
            "GET",
            f"/v1/agents/{agent_id}/memory/messages",
            params={"session_id": session_id, "limit": limit},
        )
        return ShortMemory._from(data)

    def recall(
        self,
        agent_id: str,
        query: str,
        *,
        limit: int = 10,
        min_similarity: float = 0.0,
    ) -> RecallResult:
        """Semantic search over long-term memories using the query text."""
        data = self._http.request(
            "POST",
            f"/v1/agents/{agent_id}/memory/recall",
            json={"query": query, "limit": limit, "min_similarity": min_similarity},
        )
        return RecallResult._from(data)

    def extract(
        self,
        agent_id: str,
        conversation: str,
        *,
        session_id: str | None = None,
    ) -> ExtractResult:
        """Extract facts from a conversation with Claude and persist as long-term memories."""
        data = self._http.request(
            "POST",
            f"/v1/agents/{agent_id}/memory/extract",
            json={"conversation": conversation, "session_id": session_id},
        )
        return ExtractResult._from(data)

    def list(
        self,
        agent_id: str,
        *,
        page: int = 1,
        page_size: int = 20,
    ) -> MemoryPage:
        """List all long-term memories for an agent, paginated."""
        data = self._http.request(
            "GET",
            f"/v1/agents/{agent_id}/memory",
            params={"page": page, "page_size": page_size},
        )
        return MemoryPage._from(data)

    def delete(self, agent_id: str, memory_id: str) -> None:
        """Soft-delete a long-term memory record."""
        self._http.request("DELETE", f"/v1/agents/{agent_id}/memory/{memory_id}")


class AsyncMemoryClient:
    """Asynchronous memory operations."""

    def __init__(self, http: AsyncTransport) -> None:
        self._http = http

    async def append(
        self,
        agent_id: str,
        role: str,
        content: str,
        *,
        session_id: str = "default",
        metadata: dict[str, Any] | None = None,
    ) -> ShortMemory:
        """Append a message to the agent's short-term session memory."""
        data = await self._http.request(
            "POST",
            f"/v1/agents/{agent_id}/memory/messages",
            params={"session_id": session_id},
            json={"role": role, "content": content, "metadata": metadata or {}},
        )
        return ShortMemory._from(data)

    async def messages(
        self,
        agent_id: str,
        *,
        session_id: str = "default",
        limit: int = 50,
    ) -> ShortMemory:
        """Retrieve recent messages from the agent's session memory."""
        data = await self._http.request(
            "GET",
            f"/v1/agents/{agent_id}/memory/messages",
            params={"session_id": session_id, "limit": limit},
        )
        return ShortMemory._from(data)

    async def recall(
        self,
        agent_id: str,
        query: str,
        *,
        limit: int = 10,
        min_similarity: float = 0.0,
    ) -> RecallResult:
        """Semantic search over long-term memories using the query text."""
        data = await self._http.request(
            "POST",
            f"/v1/agents/{agent_id}/memory/recall",
            json={"query": query, "limit": limit, "min_similarity": min_similarity},
        )
        return RecallResult._from(data)

    async def extract(
        self,
        agent_id: str,
        conversation: str,
        *,
        session_id: str | None = None,
    ) -> ExtractResult:
        """Extract facts from a conversation with Claude and persist as long-term memories."""
        data = await self._http.request(
            "POST",
            f"/v1/agents/{agent_id}/memory/extract",
            json={"conversation": conversation, "session_id": session_id},
        )
        return ExtractResult._from(data)

    async def list(
        self,
        agent_id: str,
        *,
        page: int = 1,
        page_size: int = 20,
    ) -> MemoryPage:
        """List all long-term memories for an agent, paginated."""
        data = await self._http.request(
            "GET",
            f"/v1/agents/{agent_id}/memory",
            params={"page": page, "page_size": page_size},
        )
        return MemoryPage._from(data)

    async def delete(self, agent_id: str, memory_id: str) -> None:
        """Soft-delete a long-term memory record."""
        await self._http.request("DELETE", f"/v1/agents/{agent_id}/memory/{memory_id}")

File: sdk-typescript/src/resources/memory.ts
import type { HttpClient } from "../http.js";
import type {
  ShortMemory,
  MemoryItem,
  RecallResult,
  ExtractResult,
  MemoryPage,
  MemoryStats,
  Role,
} from "../types.js";

export interface AppendParams {
  agentId: string;
  role: Role;
  content: string;
  sessionId?: string;
  metadata?: Record<string, unknown>;
}

export interface MessagesParams {
  agentId: string;
  sessionId: string;
}

export interface RecallParams {
  agentId: string;
  query: string;
  limit?: number;
  minSimilarity?: number;
  sessionId?: string;
  tags?: string[];
}

export interface ExtractParams {
  agentId: string;
  sessionId: string;
  /** Minimum importance score (0–1) for extracted memories. */
  minImportance?: number;
}

export interface ListMemoriesParams {
  agentId: string;
  limit?: number;
  page?: number;
  includeArchived?: boolean;
  tags?: string[];
}

export interface DeleteMemoryParams {
  agentId: string;
  memoryId: string;
}

export class MemoryResource {
  constructor(private readonly http: HttpClient) {}

  /**
   * Append a message to the short-term memory for a session.
   * Messages are stored in Redis and feed into long-term extraction.
   */
  async append(params: AppendParams): Promise<ShortMemory> {
    const { agentId, ...body } = params;
    return this.http.post<ShortMemory>(
      `/v1/agents/${agentId}/memory/messages`,
      {
        role: body.role,
        content: body.content,
        session_id: body.sessionId,
        metadata: body.metadata ?? {},
      }
    );
  }

  /** Retrieve the short-term message history for a session. */
  async messages(params: MessagesParams): Promise<ShortMemory> {
    return this.http.get<ShortMemory>(
      `/v1/agents/${params.agentId}/memory/messages`,
      { params: { session_id: params.sessionId } }
    );
  }

  /**
   * Semantic search over long-term memories using cosine similarity.
   * Returns memories sorted by relevance.
   */
  async recall(params: RecallParams): Promise<RecallResult> {
    const { agentId, query, limit, minSimilarity, sessionId, tags } = params;
    return this.http.post<RecallResult>(
      `/v1/agents/${agentId}/memory/recall`,
      {
        query,
        limit,
        min_similarity: minSimilarity,
        session_id: sessionId,
        tags,
      }
    );
  }

  /**
   * Extract and persist long-term memories from a session's message history.
   * Uses Claude to identify important facts, decisions and context.
   */
  async extract(params: ExtractParams): Promise<ExtractResult> {
    const { agentId, sessionId, minImportance } = params;
    return this.http.post<ExtractResult>(
      `/v1/agents/${agentId}/memory/extract`,
      { session_id: sessionId, min_importance: minImportance }
    );
  }

  /** List long-term memories with optional pagination and archived filter. */
  async list(params: ListMemoriesParams): Promise<MemoryPage> {
    const { agentId, limit, page, includeArchived, tags } = params;
    return this.http.get<MemoryPage>(`/v1/agents/${agentId}/memory`, {
      params: {
        limit,
        page,
        include_archived: includeArchived,
        tags: tags?.join(","),
      },
    });
  }

  /** Get aggregate memory stats for an agent (counts, avg importance, etc.). */
  async stats(agentId: string): Promise<MemoryStats> {
    return this.http.get<MemoryStats>(`/v1/agents/${agentId}/memory/stats`);
  }

  /** Hard-delete a single long-term memory entry. */
  async delete(params: DeleteMemoryParams): Promise<void> {
    await this.http.delete(`/v1/agents/${params.agentId}/memory/${params.memoryId}`);
  }
}
